discojs/src/privacy.ts

```diff
@@ -12,13 +14,85 @@ async function frobeniusNorm(weights: WeightsContainer): Promise<number> {
   return Math.sqrt(squared[0]);
 }
 
-/** Scramble weights */
-export function addNoise(
+/** Compute the frobenius norm of weight for one layer */
+export async function layerWiseFrobeniusNorm(weight: tf.Tensor): Promise<number>{
+  const squared = await weight.square().sum().data();
+  return Math.sqrt(squared[0]);
+}
+
+/** ALDP-FL implementation */
+// Conditions need to be added for the first three epochs -> get the avg update from all of the available previous updates
+export function getClippingRadius(weightHistory: WeightHistory, defaultRadius:number|undefined = 20, windowSize: number=3, minRadius=1e-12): number | number[]{
+  // When there is no update in weightHistory yet
+  if (weightHistory.size === 0) return defaultRadius ?? 20;
+
+  const radii = Array.from(weightHistory.entries()).map(([_, norms]) => {
+    if (!norms || norms.length === 0) return defaultRadius ?? 20;
+
+    if (norms.length < windowSize){
+      // average over all available entries
+      const avg = norms.reduce((sum, n) => sum+n, 0) / norms.length;
+
+      if (avg > defaultRadius)
+        return defaultRadius
+
+      return Math.max(minRadius, avg);
+    } else{
+      const recent = norms.slice(-windowSize);
+      const avg = recent.reduce((sum, n) => sum+n, 0) / recent.length;
+
+      if (avg > defaultRadius)
+        return defaultRadius
+
+      return Math.max(minRadius, avg);
+    }
```

Comment:
both branches end in the same way and pretty much do the same thing, we can simplify it (and the branching logic also)
```suggestion
    const recent = norms.slice(-windowSize);
    const avg = recent.reduce((sum, n) => sum+n, 0) / recent.length;

    return Math.max(minRadius, Math.min(avg, defaultRadius));
```